core/eth/state.go

```diff
@@ -1,82 +1,309 @@
 	Client common.EthClient
-	Tx     core.Reader
+	Reader core.Reader
+	logger logging.Logger
+	// A cache map of the operator registry, key: operator id, value: socket string
+	SocketMap map[core.OperatorID]*string
+	// Mutex to access socket map
+	socketMu sync.Mutex
+	// The previous block number the socket map was updated at, inclusive
+	socketPrevBlockNumber atomic.Uint32
 }
 
-func NewChainState(tx core.Reader, client common.EthClient) *ChainState {
-	return &ChainState{
-		Client: client,
-		Tx:     tx,
+func NewChainState(reader core.Reader, client common.EthClient, logger logging.Logger) (*ChainState, error) {
+	currentBlockNumber, err := client.BlockByNumber(context.Background(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current block number: %w", err)
+	}
+	cs := &ChainState{
+		Client:    client,
+		Reader:    reader,
+		logger:    logger,
+		SocketMap: make(map[core.OperatorID]*string),
 	}
+	// Set initial block number to current block number
+	cs.socketPrevBlockNumber.Store(uint32(currentBlockNumber.Number().Uint64()))
+	return cs, nil
 }
 
 var _ core.ChainState = (*ChainState)(nil)
 
+// GetOperatorStateByOperator returns the operator state for a given operator at a specific block number.
 func (cs *ChainState) GetOperatorStateByOperator(ctx context.Context, blockNumber uint, operator core.OperatorID) (*core.OperatorState, error) {
-	operatorsByQuorum, _, err := cs.Tx.GetOperatorStakes(ctx, operator, uint32(blockNumber))
+	operatorsByQuorum, _, err := cs.Reader.GetOperatorStakes(ctx, operator, uint32(blockNumber))
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to get operator stakes for operator %x at block %d: %w", operator, blockNumber, err)
 	}
 
-	return getOperatorState(operatorsByQuorum, uint32(blockNumber))
+	return cs.getOperatorState(ctx, operatorsByQuorum, uint32(blockNumber))
+}
 
+// GetOperatorState returns the operator state for a given quorum at a specific block number.
+func (cs *ChainState) GetOperatorState(creader context.Context, blockNumber uint, quorums []core.QuorumID) (*core.OperatorState, error) {
+	operatorsByQuorum, err := cs.Reader.GetOperatorStakesForQuorums(creader, quorums, uint32(blockNumber))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get operator stakes for quorums %v at block %d: %w", quorums, blockNumber, err)
+	}
+
+	return cs.getOperatorState(creader, operatorsByQuorum, uint32(blockNumber))
 }
 
-func (cs *ChainState) GetOperatorState(ctx context.Context, blockNumber uint, quorums []core.QuorumID) (*core.OperatorState, error) {
-	operatorsByQuorum, err := cs.Tx.GetOperatorStakesForQuorums(ctx, quorums, uint32(blockNumber))
+// GetCurrentBlockNumber returns the current block number.
+func (cs *ChainState) GetCurrentBlockNumber(creader context.Context) (uint, error) {
+	number, err := cs.Client.BlockNumber(creader)
 	if err != nil {
-		return nil, err
+		cs.logger.Warn("failed to get current block number: %w", "error", err)
```

Comment:
I notice a lot of your log statements include flags such as `%w`, but my understanding is that log doesn't do this sort of fmt replacement. I'm not 100% sure on that though, so consider checking log output in a test to confirm if you're unsure as well